core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -777,6 +787,75 @@ NavigableMap<Long, InFlightBatch> cachedState() {
+                            if (updateResult == null) {
+                                log.debug("Unable to release acquisition lock on timeout for the batch: {}"
+                                        + " for the share partition: {}-{}-{}", inFlightBatch, groupId, memberId, topicIdPartition);
+                            } else {
+                                localNextFetchOffset = Math.min(cachedStateEntry.getKey(), localNextFetchOffset);
+                            }
+                        } else {
+                            log.trace("The batch is not in acquired state while release of acquisition lock on timeout, skipping, batch: {}"
+                                            + " for the share group: {}-{}-{}", inFlightBatch, groupId, memberId, topicIdPartition);
+                        }
+                    } else { // Case when batch has a valid offset state map.
+                        for (Map.Entry<Long, InFlightState> offsetState : inFlightBatch.offsetState.entrySet()) {
+                            // For the first batch which might have offsets prior to the request base
+                            // offset i.e. cached batch of 10-14 offsets and request batch of 12-13.
+                            if (offsetState.getKey() < baseOffset) {
+                                continue;
+                            }
+
+                            if (offsetState.getKey() > lastOffset) {
+                                // No further offsets to process.
+                                break;
+                            }
+
+                            if (offsetState.getValue().state != RecordState.ACQUIRED) {
+                                log.trace("The offset is not in acquired state while release of acquisition lock on timeout, skipping, offset: {} batch: {}"
+                                                + " for the share group: {}-{}-{}", offsetState.getKey(), inFlightBatch,
+                                        groupId, memberId, topicIdPartition);
+                                continue;
+                            }
+                            InFlightState updateResult = offsetState.getValue().tryUpdateState(RecordState.AVAILABLE, false);
+                            if (updateResult == null) {
+                                log.debug("Unable to release acquisition lock on timeout for the offset: {} in batch: {}"
+                                                + " for the share group: {}-{}-{}", offsetState.getKey(), inFlightBatch,
+                                        groupId, memberId, topicIdPartition);
+                                continue;
+                            }
+                            localNextFetchOffset = Math.min(offsetState.getKey(), localNextFetchOffset);
+                        }
+                    }
+                    nextFetchOffset = localNextFetchOffset;
```

Comment:
Seems the code is similar to one above we have to acknowledge the offset and batch. Why not to have a common code which accepts to and from state transition?

Reply:
There are differences here. The only thing common would be 2 if statements at top, rest is different.